cronofy.php

```diff
@@ -405,65 +310,129 @@ function upsert_event($params)
+    }
 
-        $postfields = array('event_id' => $params['event_id']);
+    public function create_channel($params)
+    {
+        /*
+          String callback_url : The URL that is notified whenever a change is made. REQUIRED
 
-        $result = $this->http_delete($url, $postfields, $headers);
+          returns $result - Details of new channel. Details are available in the Cronofy API Documentation
+        */
+        $postfields = array('callback_url' => $params['callback_url']);
 
-        if (empty($result)) {
-            return true;
-        } else {
-            return json_decode($result, true);
-        }
+        return $this->http_post("/" . self::API_VERSION . "/channels", $postfields);
+    }
+
+    public function list_channels()
+    {
+        /*
+          returns $result - Array of channels. Details are available in the Cronofy API Documentation
+         */
+        return $this->http_get("/" . self::API_VERSION . "/channels");
     }
 
-    function delete_external_event($params)
+    public function close_channel($params)
+    {
+        /*
+          channel_id : The ID of the channel to be closed. REQUIRED
+
+          returns $result - Array of channels. Details are available in the Cronofy API Documentation
+         */
+        return $this->http_delete("/" . self::API_VERSION . "/channels/" . $params['channel_id']);
+    }
+
+    public function delete_external_event($params)
     {
         /*
           calendar_id : The calendar_id of the calendar you wish the event to be removed from. REQUIRED
           String event_uid : The String that uniquely identifies the event. REQUIRED
 
           returns true on success, associative array of errors on failure
          */
-        $url = self::API_ROOT_URL . "/" . self::API_VERSION . "/calendars/" . $params['calendar_id'] . "/events";
+        $postfields = array('event_uid' => $params['event_uid']);
+
+        return $this->http_delete("/" . self::API_VERSION . "/calendars/" . $params['calendar_id'] . "/events", $postfields);
+    }
+
+    public function authorize_with_service_account($params)
+    {
+        /*
+          email : The email of the user to be authorized. REQUIRED
+          scope : The scopes to authorize for the user. REQUIRED
+          callback_url : The URL to return to after authorization. REQUIRED
+         */
+        if (isset($params["scope"]) && gettype($params["scope"]) == "array") {
+            $params["scope"] = join(" ", $params["scope"]);
+        }
 
+        return $this->http_post("/" . self::API_VERSION . "/service_account_authorizations", $params);
+    }
+
+    private function api_url($path)
+    {
+        return self::API_ROOT_URL . $path;
+    }
+
+    private function url_params($params)
+    {
+        if (count($params) == 0) {
+            return "";
+        }
+        $str_params = array();
+
+        foreach ($params as $key => $val) {
+            array_push($str_params, $key . "=" . urlencode($val));
```

Comment:
This should handle array values differently as it should be `$key."[]=".urlencode($val)."&"` or something